pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -0,0 +1,440 @@
+		command = EncryptionKeyRotationReencrypt
+	default:
+		return plan.OneTimeInstruction{}, fmt.Errorf("cannot create secrets-encrypt instruction for node with phase: [%s]", phase)
+	}
+
+	return plan.OneTimeInstruction{
+		Name:    SecretsEncryptApplyCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			command,
+		},
+	}, nil
+}
+
+func encryptionKeyRotationStatusInstruction(cp *rkev1.RKEControlPlane) plan.OneTimeInstruction {
+	return plan.OneTimeInstruction{
+		Name:    SecretsEncryptStatusCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			"status",
+		},
+		SaveOutput: true,
+	}
+}
+
+func encryptionKeyRotationPeriodicStatusInstruction(cp *rkev1.RKEControlPlane) plan.PeriodicInstruction {
+	return plan.PeriodicInstruction{
+		Name:    SecretsEncryptStatusCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			"status",
+		},
+		PeriodSeconds: 2,
+	}
+}
+
+// encryptionKeyRotationRestartInstruction generates a restart command for the rke2/k3s server, using the last known leader phase in order to
+// ensure that non-init nodes have a refreshed plan if the leader phase changes. If secrets-encrypt commands were run on
+// a node that is not the init node, this ensures that after this situation is identified and the leader is restarted,
+// other control plane nodes will be restarted given that the leader phase will have changed without a corresponding
+// apply command.
+func encryptionKeyRotationRestartInstruction(cp *rkev1.RKEControlPlane, entry *planEntry, leaderPhase string) plan.OneTimeInstruction {
+	var lastKnownLeaderPhaseEnv string
+	switch entry.Metadata.Labels[rke2.CattleOSLabel] {
+	case windows:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("$env:LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
+	default:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
+	}
+	return plan.OneTimeInstruction{
+		Name:    "restart-service",
+		Command: "systemctl",
+		Args: []string{
+			"restart", rke2.GetRuntimeServerUnit(cp.Spec.KubernetesVersion),
+		},
+		Env: []string{
+			lastKnownLeaderPhaseEnv,
+		},
+	}
+}
+
+func encryptionKeyRotationFindApplyCommand(entry *planEntry) (plan.OneTimeInstruction, error) {
+	if entry.Plan == nil {
+		return plan.OneTimeInstruction{}, fmt.Errorf("could not find apply command for [%s]: plan is empty", entry.Machine.Name)
+	}
+	for _, i := range entry.Plan.Plan.Instructions {
+		if i.Name == SecretsEncryptApplyCommand {
+			return i, nil
+		}
+	}
+	return plan.OneTimeInstruction{}, fmt.Errorf("could not find apply command for [%s]", entry.Machine.Name)
+}
+
+// encryptionKeyRotationExpectedPhase returns the expected secrets-encrypt phase extracted from the leader's apply command present in the plan.
+func encryptionKeyRotationExpectedPhase(leaderPhase string, leader *planEntry) (bool, error) {
```

Comment:
The name of this function implies that it returns the expected status. It seems that it returns whether the phase is what is expected based on the command.
```suggestion
func encryptionKeyRotationPhaseIsExpectedFromCommand(leaderPhase string, leader *planEntry) (bool, error) {
```